Fix lazy bug

Robbepop opened the pull request:
The problem was that if a `Lazy` instance has not been used in an execution (which it its whole point of existence) then it corrupted the `KeyPtr` for all following computations.

Unfortunately this bug went unnoticed in our extensive test suite.

- [x] Fixed the code causing the issue.
- [x] Cleaned up the surrounding code.
- [x] Added regression unit test.

ascjones commented:
Is it possible to make the tests fail for this case in the offchain env?

ascjones reviewed (approved):
Confirmed fixes erc20. Still curious about whether the offchain env should catch such bugs

Robbepop commented:
> Is it possible to make the tests fail for this case in the offchain env?

I just added a regression test that fails without this fix.
Also cleaned-up implementation of `LazyCell` and the underlying `StorageEntry`.

cmichi reviewed (approved):
Thanks for adding the test!
